theano/gpuarray/opt.py

```diff
@@ -1207,7 +1207,7 @@ def local_gpua_careduce(op, context_name, inputs, outputs):
             return False
         x, = inputs
         idtype = x.dtype
-        adtype = getattr(op, 'acc_dtype', None)
+        adtype = getattr(op, 'acc_dtype', idtype)
```

Comment:
Why this change? I don't understand it.

Reply:
It's because of this line in optimization that converts gpucareduce to gpudnnreduction: https://github.com/Theano/Theano/blob/master/theano/gpuarray/dnn.py#L3737

The `acc dtype` in CPU careduce may be undefined or set to None, but it is currently updated by the optimization that converts careduce to gpucareducecuda (1) for float32 case, so that everything goes well in float32. But when floatX=float64, `acc dtype` will still be None after optimization from CPU to GPU (1), and then optimization from GPU to cuDNN will fail at line pointed above.

(1): https://github.com/Theano/Theano/blob/master/theano/gpuarray/opt.py#L1210

I detected this problem with `theano.gpuarray.tests.test_dnn.test_dnn_reduction_absmax`, which failed with floatX=float64 on my computer.

With the introduced change, I make sure `acc dtype` falls back to input dtype if it is not defined by CPU careduce.

Reply:
From tensor/elemwise.py line 1340, I think we should use the output dtype as the acc_dtype when there is no acc_dtype.

Reply:
This line (tensor/elemwise.py line 1340) is in `perform()` method, so it does not change the op attributes (and it will not be called when optimizing to GPU).

Reply:
PS: Maybe can I just fix it for float64 as it is done for float32 ?  https://github.com/Theano/Theano/blob/master/theano/gpuarray/opt.py#L1216

With float64:
```python
        if idtype == 'float64' and odtype == 'float64':
            adtype = 'float64'
```

Reply:
This won't be enough. Your change to assing_init are good. We should not always use float32 inside it. I'm still reviewing your new commit.